Initial implementation of `bndl` tool used to generate and modify ConductR bundles

longshorej opened the pull request:
This PR implements `bndl`, a tool used to create and modify ConductR bundles. It currently handles input formats of `docker` and `oci-image` in both directory format (provided via `source` argument) and tar input stream (provided via `stdin` or `source` argument) It could fairly easily be extended in the future to support format `conductr`.

It auto-detects input based on some heuristics derived from `docker save` tar stream format as well as presence of certain files in the beginning of the stream or on disk. The format detection isn't bullet-proof but seems to work well enough for now. If it can't detect, a `-f` flag exists so you can explicitly tell it the format, but this shouldn't need to be used often. 

It uses the first `ref` or `tag` it comes across if the `--tag` argument isn't specified. This is typically `latest` for streams coming from `docker save`. If need be, a specific selection can be provided through that argument, i.e. `-t latest`.

If a `-o` argument is provided, output is written to the provided file. Otherwise, output is streamed to `stdout` similar to how `shazar` now operates.

Unless a `--no-shazar` argument is provided, `bndl` outputs in an already shazared format, thus allowing it to be piped directly to `conduct load`.

It currently uses a temporary directory to assemble the bundle and a temporary file to build a zip but conceptually this could be improved in the future to immediately start writing to stdout (ala streaming) if certain conditions are met.

huntc commented:
I think I'd prefer that we auto-detect the type of input if possible, or use flags to assist. I'd rather not have sub-commands for this tool as that suggests that it can take on many functions. I think `bndl` should be quite narrow in terms of its function i.e. create a bundle, possible from an existing bundle (which may be of different types).

longshorej commented:
Okay, this PR has been updated. I updated the first post so please read that. It's ready for review so let me know what you think.

### Manual Tests / Usage
------------------------------

**docker save to bundle.zip**
```bash
$ docker save alpine:latest | bndl -o bundle.zip
-> 0

$ unzip bundle.zip 
Archive:  bundle.zip
 extracting: alpine/bundle.conf      
 extracting: alpine/oci-image/oci-layout  
 extracting: alpine/oci-image/refs/latest  
 extracting: alpine/oci-image/blobs/sha256/b0ad9c3655334ffe706f81c71d18ed20a637205b11d6aec616271905372d20e1  
 extracting: alpine/oci-image/blobs/sha256/8ff8205c4c585247f86e6dddc13402ecbef5308a77cc133562a10f0dd2723038  
 extracting: alpine/oci-image/blobs/sha256/0beb6de4d6752dab3ca8fd87adcd994180d472ba98f5ce638ab0ce8e39dc246a  
 extracting: alpine/oci-image/blobs/sha256/c3efb49a20aebfcc210219556f0c7a9242265e2d76bf31d5ce3720a1635ce63b  
-> 0

$ cat alpine/bundle.conf 
name = "alpine"
components {
  oci-image {
    description = ""
    file-system-type = "oci-image"
    start-command = [
      "ociImageTag"
      "latest"
    ]
  }
}-> 0
```

**oci image directory to bundle.zip**
```bash
$ bndl ../alpine-oci -o bundle.zip
Error: bndl: OCI Image support requires that you provide a --name argument
-> 2

$ bndl ../alpine-oci --name alpine-from-oci -o bundle.zip
-> 0

$ unzip bundle.zip 
Archive:  bundle.zip
 extracting: alpine-from-oci/bundle.conf  
 extracting: alpine-from-oci/oci-image/oci-layout  
 extracting: alpine-from-oci/oci-image/refs/latest  
 extracting: alpine-from-oci/oci-image/blobs/sha256/627beaf3eaaff1c0bc3311d60fb933c17ad04fe377e1043d9593646d8ae3bfe1  
 extracting: alpine-from-oci/oci-image/blobs/sha256/9d35e442ad87c2ba518765512afc2c84c635c7a0f68fa4d3528e262b3885b0b2  
 extracting: alpine-from-oci/oci-image/blobs/sha256/251cf767ca3706af7df7658aa011b15a2f0e1722cb2bbb6c5dc1057325f0bb0a  

$ cat alpine-from-oci/bundle.conf 
name = "alpine-from-oci"
components {
  oci-image {
    description = ""
    file-system-type = "oci-image"
    start-command = [
      "ociImageTag"
      "latest"
    ]
  }
}-> 0
```

**pipe to `conduct load` (fails because conductr doesn't understand bundle type)**
```bash
$ docker save alpine:latest | bndl | conduct load
Retrieving bundle..
Loading bundle from cache typesafe/bundle/-
Bintray credentials loaded from /home/longshorej/.lightbend/commercial.credentials
Loading bundle to ConductR..
Error: Unable to contact ConductR.                ]  11%
Error: Reason: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
Error: Start the ConductR sandbox with: sandbox run IMAGE_VERSION
-> 1
```

**pipe to `conduct load` with manual `shazar` invocation**
```bash
$ docker save alpine:latest | bndl --no-shazar | shazar | conduct load
Retrieving bundle..
Loading bundle from cache typesafe/bundle/-
Bintray credentials loaded from /home/longshorej/.lightbend/commercial.credentials
Loading bundle to ConductR..
Error: Unable to contact ConductR.                ]  11%
Error: Reason: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
Error: Start the ConductR sandbox with: sandbox run IMAGE_VERSION
-> 1
```

**showing some of the oci related files**
```bash
~/desktop/wort/alpine $ cat oci-image/oci-layout | jsl-format-json 
{
    "imageLayoutVersion": "1.0.0"
}-> 0

~/desktop/wort/alpine $ cat oci-image/refs/latest | jsl-format-json 
{
    "digest": "sha256:c3efb49a20aebfcc210219556f0c7a9242265e2d76bf31d5ce3720a1635ce63b",
    "mediaType": "application/vnd.oci.image.manifest.v1+json",
    "size": 363
}-> 0

~/desktop/wort/alpine $ cat oci-image/blobs/sha256/c3efb49a20aebfcc210219556f0c7a9242265e2d76bf31d5ce3720a1635ce63b  | jsl-format-json 
{
    "config": {
        "digest": "sha256:8ff8205c4c585247f86e6dddc13402ecbef5308a77cc133562a10f0dd2723038",
        "mediaType": "application/vnd.oci.image.config.v1+json",
        "size": 469
    },
    "layers": [
        {
            "digest": "sha256:0beb6de4d6752dab3ca8fd87adcd994180d472ba98f5ce638ab0ce8e39dc246a",
            "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
            "size": 4230300
        }
    ],
    "schemaVersion": 2
}-> 0

~/desktop/wort/alpine $ cat oci-image/blobs/sha256/8ff8205c4c585247f86e6dddc13402ecbef5308a77cc133562a10f0dd2723038 | jsl-format-json 
{
    "architecture": "amd64",
    "config": {
        "Env": [
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        ]
    },
    "created": "2017-03-03T20:32:37.723773456Z",
    "history": [
        {
            "created": "2017-03-03T20:32:37.723773456Z",
            "created_by": "/bin/sh -c #(nop) ADD file:730030a984f5f0c5dc9b15ab61da161082b5c0f6e112a9c921b42321140c3927 in / "
        }
    ],
    "os": "linux",
    "rootfs": {
        "diff_ids": [
            "sha256:23b9c7b43573dd164619ad59e9d51eda4095926729f59d5f22803bcbe9ab24c2"
        ],
        "type": "layers"
    }
}-> 0
```

**showing config from a different docker image, `hello-world` (note `Cmd` that is absent in `alpine`)**
```bash
$ docker save hello-world | bndl -o hello-world-bundle.zip

$ unzip hello-world-bundle.zip 
Archive:  hello-world-bundle.zip
 extracting: hello-world/bundle.conf  
 extracting: hello-world/oci-image/oci-layout  
 extracting: hello-world/oci-image/refs/latest  
 extracting: hello-world/oci-image/blobs/sha256/aed3378628167d1dd949df38427068b5f38c8c9da7f0c21d0f1de060de539ba2  
 extracting: hello-world/oci-image/blobs/sha256/ac8ebd7e07bf9a9ee457a65a4852f82a2fd7106f41e5767e18bd7db1975e45ea  
 extracting: hello-world/oci-image/blobs/sha256/f540b6d27f9fc59b8f91c818e26ba4a6eb6640fee88476dbfa8d73d29b5348aa  
-> 0

$ cat hello-world/oci-image/blobs/sha256/aed3378628167d1dd949df38427068b5f38c8c9da7f0c21d0f1de060de539ba2  | jsl-format-json 
{
    "architecture": "amd64",
    "config": {
        "Cmd": [
            "/hello"
        ],
        "Env": [
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        ]
    },
    "created": "2017-01-13T22:50:56.415736637Z",
    "history": [
        {
            "created": "2017-01-13T22:50:55.903893599Z",
            "created_by": "/bin/sh -c #(nop) COPY file:22b680a46dca706cf95f38d1f17b22bcaf396780b37fc67b8789dfb924616ed8 in / "
        },
        {
            "created": "2017-01-13T22:50:56.415736637Z",
            "created_by": "/bin/sh -c #(nop)  CMD [\"/hello\"]",
            "empty_layer": true
        }
    ],
    "os": "linux",
    "rootfs": {
        "diff_ids": [
            "sha256:98c944e98de8d35097100ff70a31083ec57704be0991a92c51700465e4544d08"
        ],
        "type": "layers"
    }
}-> 0
```

huntc commented on conductr_cli/bndl_main.py:
```diff
@@ -0,0 +1,597 @@
+
+    parser.add_argument('-t', '--tag',
+                        required=False,
+                        help='The name of the tag to use. For use with both docker and oci-image formats')
```
To use for what? It'd be good to enhance this message accordingly.

longshorej replied:
How about this? `The name of the tag to create a ConductR bundle from. For use with docker and oci-image formats. When absent the first tag present is used.`

longshorej replied:
This is done.

huntc commented on conductr_cli/bndl_main.py:
```diff
@@ -0,0 +1,597 @@
+
+            return image_name
+    finally:
+        shutil.rmtree(temp_dir)
```
This file feels way too large... suggest breaking it up into smaller modules.

fsat replied:
I agree.

My suggestion would be:

*Move the `bndl` method into its own `bndl.py` and rename it to `create_bundle` or something similar.*

This way it can be called from `bndl_main.py` as `bndl.create_bundle`. This will be consistent with the `conduct` and `sandbox` command where the function that's doing the actual work is encapsulated into its own module (i.e. `conduct_info` module for `conduct info` command).

*Move the Docker specific functionality into `bndl_docker` and OCI into `bndl_oci`**

Then these can be called from `bndl.create_bundle`.

longshorej replied:
Thanks for the ideas - I followed that organization pretty closely. This is done.

huntc reviewed (approved):
Looking great! Small comments.

fsat commented on conductr_cli/bndl_main.py:
```diff
@@ -0,0 +1,597 @@
+        else:
+            return None
+    except UnicodeError:
+        return None
```
I think it would be good to have a comment as to what may have caused the `UnicodeError`.

longshorej replied:
This is done - refactored that a bit as well.

fsat commented on conductr_cli/bndl_main.py:
```diff
@@ -0,0 +1,597 @@
+
+                    # TODO investigate if below is still the case
+                    # bundles are packaged into zips, so we don't want to compress, but OCI tooling
+                    # as of 2017-03-14 is broken for plain tar files; they must be gzip
```
Is this `TODO` still need to be investigated?

longshorej replied:
Yeah, this will need to be investigated some time in several months when OCI tooling has matured, specifically `umoci` and `oci-create-runtime-bundle`. With that caveat, what's the best action to take? I see one of A) leave as-is, B) leave as-is, but also open an issue, and C) remove comment, open an issue.

fsat reviewed (changes requested):
My feedback would be primarily splitting the `bndl.py` into smaller modules.

longshorej commented:
Thanks for the feedback! I agree and will break it up into smaller modules.

longshorej commented:
I've addressed PR feedback. Note that the PR remains a WIP because I want to rerun the manual testing but I would appreciate another review now that it's been split up into smaller components. Thanks!

longshorej commented:
Manual tests completed. Output is unchanged from above so I'm not including it here. @fsat Could you do another review on this?

longshorej committed:
Initial implementation of "bndl" tool used to generate and modify ConuctR bundles
